nomad/fsm.go

```diff
@@ -565,14 +572,30 @@ func (n *nomadFSM) applyDeregisterJob(buf []byte, index uint64) interface{} {
 		panic(fmt.Errorf("failed to decode request: %v", err))
 	}
 
-	return n.state.WithWriteTransaction(func(tx state.Txn) error {
-		if err := n.handleJobDeregister(index, req.JobID, req.Namespace, req.Purge, tx); err != nil {
+	err := n.state.WithWriteTransaction(func(tx state.Txn) error {
+		err := n.handleJobDeregister(index, req.JobID, req.Namespace, req.Purge, tx)
+
+		if err != nil {
 			n.logger.Error("deregistering job failed", "error", err)
 			return err
 		}
 
 		return nil
 	})
+
+	// always attempt upsert eval even if job deregister fail
+	if req.Eval != nil {
+		req.Eval.JobModifyIndex = index
+		if err := n.upsertEvals(index, []*structs.Evaluation{req.Eval}); err != nil {
+			return err
+		}
+	}
```

Comment:
Note that here we attempt to insert the eval even if the job deregistration fails.  I find this behavior very odd but it's explicitly tested in https://github.com/hashicorp/nomad/blob/97c69ee9a7312aaa315fec5850fe0a58e13a238a/nomad/job_endpoint_test.go#L2969 - with it insuring that the resulting eval is populated and have JobModifyIndex set ?!  I kept the behavior the same

Reply:
Maybe this was intended to protect against the previous non-atomic behavior? Ex. if you deregister but the eval wasn't persisted, but then tried to deregister again, the job would be potentially gone but you'd still want an eval to clean it up?

Reply:
This actually seems like a bug to me, but I think one that does not impact correctness (only waste some work doing nothing).

From peeking around `generic_sched.go` and `reconcile.go`, I'm not seeing us check if the evaluation is a Deregister. *We always seem to checked if Job.Stopped is `true`!* This means if we fail to update the statestore with the stopped Job, the Deregister evaluation will be the same as any other evaluation and end up being a noop for the job as it is not stopped and presumably already scheduled/allocated.

I suspect that @tgross is correct and that the test is merely asserting the non-atomic behavior existed: https://github.com/hashicorp/nomad/pull/981

Since we're making Job+Eval submissions atomic, I think we should at least trying to make this section atomic as well. I can't figure out where a Deregister eval for a non-Stopped job would have a desirable affect, but perhaps somebody else can find a case?

Reply:
I'm inclined to let the current behavior stand as-is, as stopping making an eval is a user visible change.  Will follow up in another PR and we can discuss this issue further there.

Reply:
> I'm inclined to let the current behavior stand as-is, as stopping making an eval is a user visible change.

I don't find this reason sufficient to keep it since the eval in question is a `dereg` but the effect would be a noop. If that's the behavior I feel like we're emitting a useless and actively misleading eval which should be treated like a bug and removed. There's no benefit to leaving it in place as anyone observing it would only be confused by its lack of affect.

Reply:
I agree that it should be changed - but I believe such a user-visible behavior (albeit a small one) change is outside the scope of this PR, so I will follow up in another PR.